File: October_2/Solution.cpp
class Solution {
public:
int dp[31][1001] = {0};
int numRollsToTarget(int n, int k, int target) {
dp[0][0] = 1;

    for(int i = 1;i<=target;i++){
        dp[0][i] = 0;
    }
    for(int i=1;i<=n;i++){
        dp[i][0] = 0;
    }
    for(int i=1;i<=n;i++){
        for(int j=1;j<=target;j++){
            long long res = 0;
            for(int x=1;x<=k;x++){
                if(j>=x){
                    res += dp[i-1][j-x] % 1000000007;
                    res = res % 1000000007;
                }
            }
            dp[i][j] = res ;
        }
    }
    return dp[n][target] ;
}};

File: October_2/solution.py
class Solution:             # Here's the plan:
                            #   We apply dp by recursion, in which we define the simpler problem as determining the
                            #   result for each possible roll of the ith die. The state tree is complete n-nary k levels,
                            #   which has n^k-1 nodes, but the recursive call has only n*k possible inputs, so caching
                            #   the recursive function call reduces the number of computations significantly.
    @lru_cache(maxsize=None)
    def numRollsToTarget(self, n: int, k: int, target: int) -> int:

        if target <= 0 or n*k < target :
            return 0

        if n==1:
            return int(target <= k)

        for i in range(1, k+1):
            ans = sum(self.numRollsToTarget(n-1,k,target-i)
                     for i in range(1, k+1))
        
        return ans % (10**9+7)
